bin/proving-service/src/api/mod.rs

```diff
@@ -21,48 +24,133 @@ pub struct RpcListener {
+    ) -> Result<Response<ProvingResponse>, tonic::Status> {
         let prover = self
-            .local_prover
+            .provers
             .try_lock()
             .map_err(|_| Status::resource_exhausted("Server is busy handling another request"))?;
 
-        let transaction_witness =
-            TransactionWitness::read_from_bytes(&request.get_ref().transaction_witness)
-                .map_err(invalid_argument)?;
+        let prover = prover
+            .tx_prover
+            .as_ref()
+            .ok_or(Status::unimplemented("Transaction prover is not enabled"))?;
+
+        let transaction_witness = TransactionWitness::read_from_bytes(&request.get_ref().payload)
+            .map_err(invalid_argument)?;
 
         let proof = prover.prove(transaction_witness).map_err(internal_error)?;
 
         // Record the transaction_id in the current tracing span
         let transaction_id = proof.id();
-        tracing::Span::current().record("transaction_id", tracing::field::display(&transaction_id));
+        tracing::Span::current().record("id", tracing::field::display(&transaction_id));
+
+        Ok(Response::new(ProvingResponse { payload: proof.to_bytes() }))
+    }
 
-        Ok(Response::new(ProveTransactionResponse { proven_transaction: proof.to_bytes() }))
+    #[instrument(
+        target = MIDEN_PROVING_SERVICE,
+        name = "proving_service:prove_batch",
+        skip_all,
+        ret(level = "debug"),
+        fields(id = tracing::field::Empty),
+        err
+    )]
+    pub fn prove_batch(
+        &self,
+        request: Request<ProvingRequest>,
+    ) -> Result<Response<ProvingResponse>, tonic::Status> {
+        let prover = self
+            .provers
+            .try_lock()
+            .map_err(|_| Status::resource_exhausted("Server is busy handling another request"))?;
+
+        let prover = prover
+            .batch_prover
+            .as_ref()
+            .ok_or(Status::unimplemented("Batch prover is not enabled"))?;
+
+        let batch =
+            ProposedBatch::read_from_bytes(&request.get_ref().payload).map_err(invalid_argument)?;
+
+        let proof = prover.prove(batch).map_err(internal_error)?;
+
+        // Record the batch_id in the current tracing span
+        let batch_id = proof.id();
+        tracing::Span::current().record("id", tracing::field::display(&batch_id));
+
+        Ok(Response::new(ProvingResponse { payload: proof.to_bytes() }))
+    }
+}
+
+#[async_trait::async_trait]
+impl ProverApi for ProverRpcApi {
+    #[instrument(
+        target = MIDEN_PROVING_SERVICE,
+        name = "proving_service:prove",
+        skip_all,
+        ret(level = "debug"),
+        fields(id = tracing::field::Empty),
+        err
+    )]
+    async fn prove(
+        &self,
+        request: Request<ProvingRequest>,
+    ) -> Result<Response<ProvingResponse>, tonic::Status> {
+        match request.get_ref().proof_type {
+            0 => self.prove_tx(request),
```

Comment:
nit: For a type-safer approach, you could make `prove_tx()` take `TransactionWitness` and `prove_batch()` take `ProposedBatch`, and deserialize before calling each function.